Update exception handling support to current proposal

takikawa opened the pull request:
This PR updates the support of exception handling to the latest proposal (that is compatible with future 2-phase exception handling) described in https://github.com/WebAssembly/exception-handling/pull/137 and https://github.com/WebAssembly/exception-handling/pull/143.

* Adds back tagged `catch $e`, `catch_all`, and `rethrow N` from a previous version of wabt, but with updates to match the current spec (e.g., `catch_all` shares an opcode with `else`, `rethrow`'s depth indexes only catch blocks, etc).
* Adds `unwind` and `delegate` instructions.
* Removes `exnref` and `br_on_exn`.
* Updates relevant tests.

There are some details that could still change (e.g., maybe how `delegate`'s depth is validated), but I'd be happy to submit further PRs if the spec details change.

binji commented on test/dump/try-delegate.txt:
```diff
@@ -0,0 +1,99 @@
+ 000020: 06 7f                      | try i32
+ 000022: 06 40                      |   try
+ 000024: 01                         |     nop
+ 000025: 18 01                      |     delegate 1
```
looks like delegate should dedent here?

takikawa replied:
Thanks for catching this (& `unwind` too)! Latest commit should fix this issue.

binji commented on test/dump/try-unwind.txt:
```diff
@@ -55,41 +54,37 @@
+ 000020: 06 7f                      | try i32
+ 000022: 01                         |   nop
+ 000023: 41 07                      |   i32.const 7
+ 000025: 0a                         |   unwind
```
unwind should dedent too, then reindent like `catch` or `else`

binji reviewed (approved):
lgtm, thanks for doing this! Sorry for the long delay on review

dschuff commented:
I just noticed (after the discussion in #1604, thanks for pointing it out) that this PR doesn't actually solve the problem that made me notice the opcode problem, (because I was only using the disassembler). Namely `wasm-objdump` prints opcode `0x5` as `else` rather than `catch-all` even in the context of try. Presumably it's just going directly from `opcode.def`, (which doesn't have `catch_all` in this PR). But maybe let's not fix this until we decide whether we want to keep the opcodes as they are or change them.

aheejin commented:
By the way can I merge this? If the opcode is the issue we can fix it later.

takikawa commented:
@aheejin Sure, no issues on my end with having it merged. AFAIK I've addressed all the comments in review. Thanks!

I'd also be happy to submit a PR afterwards if the opcode changes.
